### PR TITLE
Add lyrics search feature 

### PR DESCRIPTION
**Pull request by lakshaykalbhor**
I have tried adding a feature that returns the lyrics of a song. Lyrics are fetched from [Genius](https://docs.genius.com/). API key will be required.

Example queries : 
- Lyrics for song 'Wish you were here'
- Lyrics for 'Hey Jude'

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -60,3 +60,6 @@ target/
 
 #Ipython Notebook
 .ipynb_checkpoints
+
+#Trash files in Mac
+.DS_Store
```
Comment (swapagarwal) [inline .gitignore]: Normally, that would go in a global .gitignore file. Let's keep only project-specific ignores here.

Comment (swapagarwal) [inline .gitignore]: Normally, that would go in a global .gitignore file. Let's keep only project-specific ignores here.

File: modules/src/__init__.py [hunk 2]
```diff
@@ -21,6 +22,7 @@
     'weather',
     'wiki',
     'xkcd',
+
```
Comment (swapagarwal) [inline modules/src/__init__.py]: remove extra line

File: modules/src/lyrics.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import requests
+import config
+from bs4 import BeautifulSoup
+from templates.generic import *
+from templates.text import TextTemplate
+import os
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: keep imports together, so `import os` would come after `import config`

File: modules/src/lyrics.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+GENIUS_KEY = os.environ.get('GENIUS_API_KEY', config.GENIUS_API_KEY)
+
+def process(input, entities):
+    output = {}
+    try:
+
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: remove extra line

File: modules/src/lyrics.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import os
+
+GENIUS_KEY = os.environ.get('GENIUS_API_KEY', config.GENIUS_API_KEY)
+
+def process(input, entities):
+    output = {}
+    try:
+            
+        query = entities['lyrics'][0]['value']
+        base_url = "http://api.genius.com"
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: use single quotes everywhere for consistency

File: modules/src/lyrics.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        headers = {'Authorization': 'Bearer %s' %(GENIUS_KEY)} 
+        search_url = base_url + "/search"
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: single quotes

File: modules/src/lyrics.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+GENIUS_KEY = os.environ.get('GENIUS_API_KEY', config.GENIUS_API_KEY)
+
+def process(input, entities):
+    output = {}
+    try:
+            
+        query = entities['lyrics'][0]['value']
+        base_url = "http://api.genius.com"
+        headers = {'Authorization': 'Bearer %s' %(GENIUS_KEY)}
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: space after %

File: modules/src/lyrics.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    try:
+            
+        query = entities['lyrics'][0]['value']
+        base_url = "http://api.genius.com"
+        headers = {'Authorization': 'Bearer %s' %(GENIUS_KEY)} 
+        search_url = base_url + "/search"
+
+        r = requests.get(search_url, data={'q': query}, headers=headers)
+        r = r.json()
+        song_api_path = r["response"]["hits"][0]["result"]["api_path"]
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: single quotes

File: modules/src/lyrics.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        search_url = base_url + "/search"
+
+        r = requests.get(search_url, data={'q': query}, headers=headers)
+        r = r.json()
+        song_api_path = r["response"]["hits"][0]["result"]["api_path"] 
+
+        song_url = base_url + song_api_path
+        r = requests.get(song_url, headers=headers)
+        r = r.json()
+        path = r["response"]["song"]["path"]
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: single quotes

File: modules/src/lyrics.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+    except:
+        error_message = 'There was some error while retrieving data from genius.com'
+        error_message += '\n Please ask me somrthing else, like:'
+        error_message += '\n Lyrics for the song Wish you were here'
+        output['error_msg'] = TextTemplate(error_message).get_message()
+        output['success'] = False
+        
+    return output
+
```
Comment (swapagarwal) [inline modules/src/lyrics.py]: remove extra lines

File: modules/tests/test_lyrics.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import modules
+
+def test_lyrics():
+    assert('lyrics' == modules.process_query("Lyrics for the song 'Wish you were here' ")[0])
+    assert('lyrics' == modules.process_query("lyrics for 'Go Robot' ")[0])
+    assert('lyrics' == modules.process_query("lyrics for the song Strawberry Fields Forever ")[0])
+    assert('music' != modules.process_query("something random")[0])
```
Comment (swapagarwal) [inline modules/tests/test_lyrics.py]: typo, music -> lyrics

**Comment by swapagarwal** [comment]
I didn't review the part where you're scraping the site for lyrics. I don't think that's a good idea unless they provide it at an endpoint. Instead, you can add a button linking to that page, and show a subset of lyrics that's returned by the API.

**Comment by lakshaykalbhor** [comment]
Genius.com doesn't directly provide any lyrics in its API, but it provides a url containing the lyrics. If scraping lyrics from that url is a no no, could I just provide the url containing the lyrics page as a button maybe?

**Comment by swapagarwal** [comment]
Oh, I was under the impression that we only get a subset of lyrics in response.
Having atleast some part of the lyrics would be great! See https://developer.musixmatch.com/documentation

**Comment by lakshaykalbhor** [comment]
From what I've seen, MusixMatch requires the id of a track to fetch information on the track. There doesn't seem to be a way to search the id of a track with a track name. Is scraping from Genius too big of a red flag? I've been using the method to fetch lyrics in one of my [projects](https://github.com/lakshaykalbhor/MusicRepair) and it seems to work fine.

**Comment by lakshaykalbhor** [comment]
I have added support for MusixMatch. It returns a subset of the lyrics (The amount that MusixMatch allows) and provides a url for viewing the entire lyrics.

**Comment by swapagarwal** [comment]
Thanks @lakshaykalbhor! 👍

**Pull request merged**